Add SPI to create webarchive for selected frames

szewai opened the pull request:
#### 8704cb8e625752d97eafa091e1c3631a5813a7ab
<pre>
Add SPI to create webarchive for selected frames
<a href="https://rdar.apple.com/160812460">rdar://160812460</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=299316">https://bugs.webkit.org/show_bug.cgi?id=299316</a>

Reviewed by Ryosuke Niwa.

This allows clients to only keep selected frames in the saved webarchive data.

API Test: SiteIsolation.CreateWebArchiveForFrames

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _createWebArchiveForFrame:completionHandler:]):
(-[WKWebView _createWebArchiveForFrames:rootFrame:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getWebArchiveDataWithFrame):
(WebKit::WebPageProxy::getWebArchiveDataWithSelectedFrames):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveForFrames)):

Canonical link: <a href="https://commits.webkit.org/300468@main">https://commits.webkit.org/300468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b5f5afb1a79104ddeb6b1fc2681484e25488b9e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122571 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42279 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32964 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129180 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74672 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/097b926c-09de-4a28-8fd2-867ef8c829a8) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124447 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50872 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93174 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61886 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9a1be3bc-f672-4420-a15b-c0f60f08f904) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109751 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73820 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ce91036f-8e33-4b71-b45e-a5cbb133140c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33279 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27908 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72667 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103963 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131906 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49512 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37689 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101706 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49887 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105971 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101574 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25809 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46951 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25104 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46282 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49370 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55119 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48838 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52189 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50519 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5373,6 +5373,41 @@ - (void)_createWebArchiveForFrame:(WKFrameInfo *)frame completionHandler:(void (
     });
 }
 
+- (void)_createWebArchiveForFrames:(NSArray<WKFrameInfo *> *)frames rootFrame:(WKFrameInfo *)rootFrame completionHandler:(void (^)(NSData *, NSError *))completionHandler
```
I guess we can also take `NSSet` but `NSArray` is more convenient due to the builtin syntax?

szewai replied:
Yes, I think `NSArray` is a more commonly choice for API; and we need to implement isEqual and hash functions for custom Objects to use `NSSet`.

rniwa commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5373,6 +5373,41 @@ - (void)_createWebArchiveForFrame:(WKFrameInfo *)frame completionHandler:(void (
+
+    HashSet<WebCore::FrameIdentifier> targetFrameIDs;
+    for (WKFrameInfo *frame in frames) {
+        if (frame && frame._handle && frame._handle->_frameHandle->frameID())
```
Can we store frame._handle in a local variable instead of accessing it 3 different times?

szewai committed:
Add SPI to create webarchive for selected frames
rdar://160812460
https://bugs.webkit.org/show_bug.cgi?id=299316

Reviewed by Ryosuke Niwa.

This allows clients to only keep selected frames in the saved webarchive data.

API Test: SiteIsolation.CreateWebArchiveForFrames

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _createWebArchiveForFrame:completionHandler:]):
(-[WKWebView _createWebArchiveForFrames:rootFrame:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getWebArchiveDataWithFrame):
(WebKit::WebPageProxy::getWebArchiveDataWithSelectedFrames):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveForFrames)):

Canonical link: https://commits.webkit.org/300468@main
